### PR TITLE
vscode-extensions.ms-vscode-cpptools: 1.20.5 -> 1.21.4

### PR DESCRIPTION
**Pull request by rowanG077**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kjeremy** [comment]
Why change the auto patching?

**Comment by rowanG077** [comment]
> Why change the auto patching?

`cpptools` and friends are statically linked now which `autoPatchElfHook` deals with normally. There is no reason to manually patch them anymore.

**Comment by kjeremy** [comment]
All three of those executables are segfaulting me:
```
[nix-shell:~/.cache/nixpkgs-review/pr-330165]$ ./results/vscode-extensions.ms-vscode.cpptools/share/vscode/extensions/ms-vscode.cpptools/bin/cpptools-wordexp 
Segmentation fault (core dumped)

[nix-shell:~/.cache/nixpkgs-review/pr-330165]$ gdb ./results/vscode-extensions.ms-vscode.cpptools/share/vscode/extensions/ms-vscode.cpptools/bin/cpptools-srv 
GNU gdb (GDB) 14.2
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./results/vscode-extensions.ms-vscode.cpptools/share/vscode/extensions/ms-vscode.cpptools/bin/cpptools-srv...
(gdb) r
Starting program: /nix/store/ls9b3p24fq9mwmmk37372dwf5kis75zg-vscode-extension-ms-vscode-cpptools-1.21.4/share/vscode/extensions/ms-vscode.cpptools/bin/cpptools-srv 
Downloading separate debug info for system-supplied DSO at 0x7ffff7fc5000
warning: Unable to find libthread_db matching inferior's thread library, thread debugging will not be available.                                                

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7f6dcec in do_init_fini ()
   from /nix/store/ls9b3p24fq9mwmmk37372dwf5kis75zg-vscode-extension-ms-vscode-cpptools-1.21.4/share/vscode/extensions/ms-vscode.cpptools/bin/libc.so
(gdb) bt
#0  0x00007ffff7f6dcec in do_init_fini ()
   from /nix/store/ls9b3p24fq9mwmmk37372dwf5kis75zg-vscode-extension-ms-vscode-cpptools-1.21.4/share/vscode/extensions/ms-vscode.cpptools/bin/libc.so
#1  0x00007ffff7f6fc54 in __libc_start_init ()
   from /nix/store/ls9b3p24fq9mwmmk37372dwf5kis75zg-vscode-extension-ms-vscode-cpptools-1.21.4/share/vscode/extensions/ms-vscode.cpptools/bin/libc.so
#2  0x00007ffff7f31d4a in libc_start_main_stage2 ()
   from /nix/store/ls9b3p24fq9mwmmk37372dwf5kis75zg-vscode-extension-ms-vscode-cpptools-1.21.4/share/vscode/extensions/ms-vscode.cpptools/bin/libc.so
#3  0x0000000000642b04 in _start ()
```

**Comment by kjeremy** [comment]
> Hmm I don't have x86 machine to test. Perhaps it's statically linked on aarch64 but not on x86.

That's my guess. We may need to conditionalize the patching based on architecture.

**Comment by rowanG077** [comment]
I have done so. See the changes now.

File: pkgs/applications/editors/vscode/extensions/ms-vscode.cpptools/default.nix [hunk 1]
```diff
@@ -40,13 +40,14 @@
 
 let
   gdbDefaultsTo = if gdbUseFixed then "${gdb}/bin/gdb" else "gdb";
+  isx86Linux = stdenv.system == "x86_64-linux";
```
Comment (kjeremy) [inline pkgs/applications/editors/vscode/extensions/ms-vscode.cpptools/default.nix:43]: I think this should be `stdenv.hostPlatform.system`

**Comment by kjeremy** [comment]
For some reason now I'm getting a permission denied error when running the executables even though they're +x.

File: pkgs/applications/editors/vscode/extensions/ms-vscode.cpptools/default.nix [hunk 2]
```diff
@@ -73,35 +76,40 @@ vscode-utils.buildVscodeMarketplaceExtension {
...
+      rm -rf ./LLVM
+      mkdir "./LLVM/"
+      find "${clang-tools}" -mindepth 1 -maxdepth 1 | xargs ln -s -t "./LLVM"
+
+      # Patching binaries
+      chmod +x bin/cpptools bin/cpptools-srv bin/cpptools-wordexp debugAdapters/bin/OpenDebugAD7
+      patchelf --replace-needed liblttng-ust.so.0 liblttng-ust.so.1 ./debugAdapters/bin/libcoreclrtraceptprovider.so
+    ''
+    + lib.optionalString isx86Linux ''
+      chmod +x bin/cpptools
```
Comment (kjeremy) [inline pkgs/applications/editors/vscode/extensions/ms-vscode.cpptools/default.nix:105]: It looks like this happens above on line 101

Comment (rowanG077) [reply pkgs/applications/editors/vscode/extensions/ms-vscode.cpptools/default.nix]: @kjeremy My bad it should have been `bin/libc.so`, should be fixed now.

**Review by kjeremy — APPROVED** [review]

**Pull request merged**